### PR TITLE
Checkout: Applied coupon should be removed when unavailable

### PR DESCRIPTION
**Pull request by taggon**
#### Changes proposed in this Pull Request

* This PR fixes the bug that doesn't show the coupon input when a coupon gets unavailable while it is applied to the cart. The wrong message is also fixed.
Checkout_—_WordPress_com

#### Testing instructions

* Apply a coupon to a paid plan in the cart on the checkout page.
  Checkout_—_WordPress_com
* **IMPORTANT!** Remove the coupon data from the server-side or get it expired.
* Switch to a variant product.
  Checkout_—_WordPress_com
* The coupon should be removed from the cart.
  Checkout_—_WordPress_com

### CODE REVIEW

**Comment by semihakocer** [comment]
@taggon what happens if the user switches back to the first plan they selected? Does the coupon get re-added?

**Comment by DavidRothstein** [comment]
Although this is an improvement, the user experience still seems pretty bad -- I'm not sure it's good enough to start deliberately introducing cases where this will happen (i.e., coupons that only work for certain billing terms).

To answer the question @semihakocer asked, it seems like the coupon is not added back when you switch back -- and also there's no warning when you switch in the first place.  It seems like ideally there'd be a confirmation that popped up when you make the initial switch, telling you that the coupon will be removed due to your changes (similar to what happens when you delete a coupon from your cart) although that might be hard to implement.

This would benefit from some design input, maybe @fditrapani or someone else.

See also for context: pau2Xa-2lC-p2 (internal discussion)

**Comment by fditrapani** [comment]
> (i.e., coupons that only work for certain billing terms).

Correct me if I'm wrong but this doesn't seem to be the purpose of this PR. 

> and also there's no warning when you switch in the first place...  It seems like ideally there'd be a confirmation that popped up

Your instinct is right here @DavidRothstein. We should likely show a notice if something changed during the process of the checkout. If the coupon becomes unavailable between the time it is applied and the person checkouts out, we should show a modal explaining that the coupon does not apply anymore and remove it from the cart. As a reference, you can see how a modal looks, and works, by removing a product from the cart. 

To take this a step further, if we do want to apply it in cases where the the coupon only applies to one term, then in that case, we should show the modal but. not remove the coupon. Instead, we would show the coupon value as zero because the person might want to switch back to the previous term.

**Comment by DavidRothstein** [comment]
>Correct me if I'm wrong but this doesn't seem to be the purpose of this PR.

Not directly, but it's part of introducing that feature, and currently there is no other PR :smiley:  See the internal discussion I linked to above and in particular pau2Xa-2lC-p2#comment-8916

I do think this PR could go in on its own, though, because on its own it does work as a bug fix (and it clearly makes things better than they were before).

File: client/my-sites/checkout/composite-checkout/composite-checkout.tsx [hunk 1]
```diff
@@ -263,6 +263,13 @@ export default function CompositeCheckout( {
 		showAddCouponSuccessMessage
 	);
 
+	// Remove the coupon when it doesn't exist on the server-side
+	useEffect( () => {
+		if ( couponStatus === 'applied' && ! couponItem ) {
```
Comment (sirbrillig) [inline client/my-sites/checkout/composite-checkout/composite-checkout.tsx]: This works, but as there's some implicit logic happening in `translateResponseCartToWPCOMCart` that could potentially break this in the future if it is changed, perhaps we should be explicit here:

```suggestion
		if ( couponStatus === 'applied' && responseCart.coupon_savings_total_integer === 0 ) {
```

(It seems to me that the latter condition is sufficient, but if you like you could also check for `! responseCart.coupon`.)

File: client/my-sites/checkout/composite-checkout/hooks/use-show-add-coupon-success-message.ts [hunk 2]
```diff
@@ -10,7 +10,7 @@ export default function useShowAddCouponSuccessMessage(
 ): void {
 	const lastCouponCode = useRef< string >( '' );
 	useEffect( () => {
-		if ( didAddCoupon && couponCode !== lastCouponCode.current ) {
+		if ( didAddCoupon && couponCode && couponCode !== lastCouponCode.current ) {
```
Comment (sirbrillig) [inline client/my-sites/checkout/composite-checkout/hooks/use-show-add-coupon-success-message.ts]: Excellent change.

**Review by sirbrillig — COMMENTED** [review]
> We should likely show a notice if something changed during the process of the checkout. If the coupon becomes unavailable between the time it is applied and the person checkouts out, we should show a modal explaining that the coupon does not apply anymore and remove it from the cart.

I agree. However, since coupon success/failure and the associated messages currently come entirely from the shopping-cart endpoint, I think we should utilize that resource in this case. The front-end shouldn't be responsible for trying to divine why the server has done something. If there's agreement, we can create an issue to track that specifically.

> I do think this PR could go in on its own, though, because on its own it does work as a bug fix (and it clearly makes things better than they were before).

I agree with this as well. My only concern is that this fix will only apply to composite checkout; not other components (eg: `PopoverCart`) that utilize the shopping cart. I'd prefer something like this to live higher up the chain, such as in the shopping cart package itself, but I'm not exactly sure where.

https://github.com/Automattic/wp-calypso/blob/2c2bc61d54583cae90bd926f3628f130dcb6891f/packages/shopping-cart/src/use-shopping-cart-reducer.ts#L258

The above code is responsible for setting the coupon status to `applied`, and it comes directly from the `responseCart.is_coupon_applied` property, which is set by the server. So the real fix here should probably be `is_coupon_applied` on the shopping-cart endpoint (created an issue for that in 68-gh-Automattic/payments-shilling). In the meantime, we could modify `getUpdatedCouponStatus` to return a different status if `is_coupon_applied` is true but `coupon_savings_total_integer` is 0. What do you think?

**Comment by sirbrillig** [comment]
> The above code is responsible for setting the coupon status to applied, and it comes directly from the responseCart.is_coupon_applied property, which is set by the server. .... we could modify getUpdatedCouponStatus to return a different status if is_coupon_applied is true but coupon_savings_total_integer is 0. What do you think?

Now that I look at `getUpdatedCouponStatus` some more, I think that this was actually close to the original intent of `couponStatus`; to identify when a coupon had been applied but had no effect. I think this deserves some refactoring in general, just as I did recently with `cacheStatus`. I have a PR in progress that I'll mention here to see if it also resolves this issue.

**Comment by sirbrillig** [comment]
It turns out that there's some major issues with the way coupons on the server were being marked as applied which we shouldn't have to deal with on the client. D53596-code should fix the server so that `is_coupon_applied` returns the correct value and when combined with https://github.com/Automattic/wp-calypso/pull/47943, it should fix the problem described in this PR without needing to remove invalid coupons.

**Comment by taggon** [comment]
@sirbrillig Thanks. I've updated the code as per your request.
@fditrapani @DavidRothstein Thanks for the feedback. A follow-up PR will be here to improve the user experience.

**Comment by sirbrillig** [comment]
Thank you, Taggon! I _think_ (and please correct me if I'm wrong) that this change may no longer be necessary.

The idea here is to remove a coupon from the cart if one is applied that has no effect. That's a great improvement! However, I stole that idea a bit and made it apply to the server response in D53596-code because it seems that this was the original intention of the `is_coupon_applied` property. It had become broken over time, and now it hopefully is fixed.

In addition, the server response was supposed to include a success or error message related to the current state of the coupon, but it was buggy and so we had this code in calypso meant to display a better message. The above patch also fixed those messages, so I was able to [remove the calypso code entirely](https://github.com/Automattic/wp-calypso/pull/47943) (you'll notice that this rebased PR now seems to add a hook which is not used anywhere). Those messages can be improved, certainly, and the improvement can be either on the server (see the `Store_Shopping_Cart`), or on the client by adding conditions to `CartMessages` here (which also allows including markup in the message):

https://github.com/Automattic/wp-calypso/blob/b9caa7dc132c9db302d010f41a984e7bef4b5e51/client/my-sites/checkout/cart/cart-messages.jsx#L69-L86

**Comment by taggon** [comment]
Rebased

File: client/my-sites/checkout/composite-checkout/hooks/use-show-add-coupon-success-message.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+/**
+ * External dependencies
+ */
+import { useEffect, useRef } from 'react';
+
+export default function useShowAddCouponSuccessMessage(
```
Comment (southp) [inline client/my-sites/checkout/composite-checkout/hooks/use-show-add-coupon-success-message.ts]: I can't find any place using this hook. Maybe it has been deprecated at some points? or maybe I'm simply overlooking?

**Review by southp — CHANGES_REQUESTED** [review]
@sirbrillig  I think this is still relevant since it will fix the issue that the coupon field disappears after switching terms. 

e.g. on production:
Kapture 2021-01-27 at 17 28 31

on this branch:
Kapture 2021-01-27 at 17 13 28
The behavior is not entirely consistent sometimes; note how the coupon input field could be either staying or disappearing. But it is indeed an improvement.

@taggon The only concern I have is that I can't locate any place using the `useShowAddCouponSuccessMessage` hook. It could be me overlooking of course. I'm happy to revisit this PR immediately once we resolve this :)

File: client/my-sites/checkout/composite-checkout/composite-checkout.tsx [hunk 4]
```diff
@@ -233,6 +233,14 @@ export default function CompositeCheckout( {
 		isInitialCartLoading,
 	} );
 
+	// Remove the coupon when it doesn't exist on the server-side
+	const couponSavingsTotal = responseCart?.coupon_savings_total_integer;
+	useEffect( () => {
```
Comment (sirbrillig) [inline client/my-sites/checkout/composite-checkout/composite-checkout.tsx]: Hm... I still think this is covering over a problem that's more fundamental. If the coupon no longer applies to the updated shopping cart, I'd expect the shopping cart endpoint to remove the coupon (or at least treat it as invalid and report a warning). I'll take a look at the bug and see if I can understand what's going on.

Comment (sirbrillig) [reply client/my-sites/checkout/composite-checkout/composite-checkout.tsx]: I reproduced this issue and the cart response actually _does_ have the coupon removed. The bug is that `getUpdatedCouponStatus` does not allow a coupon to stop being applied even if the cart says it is:

https://github.com/Automattic/wp-calypso/blob/bafff16e004276e9de18ca7981d25ed670d84e4a/packages/shopping-cart/src/use-shopping-cart-reducer.ts#L263

**Comment by sirbrillig** [comment]
@southp thanks for those screenshots! Very helpful. I found the bug, which has to do with the `shopping-cart` reducer, and not with the endpoint. This should fix the issue, I think: https://github.com/Automattic/wp-calypso/pull/49371

File: client/my-sites/checkout/composite-checkout/composite-checkout.tsx (hunk 4 above)
Comment (taggon) [reply client/my-sites/checkout/composite-checkout/composite-checkout.tsx]: Thanks @sirbrillig! The `getUpdatedCouponStatus` now returns `invalid` when the applied coupon is invalid for the current product. Even better, the code change is more compact! 👍

**Review by southp — APPROVED** [review]
After retesting, it works as expected and the code is even more concise. Nice :)
Since it's a subset of what @sirbrillig is working on so it shouldn't cause too much headache, I recommend us to ship this first to mitigate this UX issue sooner than later.

**Comment by taggon** [comment]
This PR is closed in favor of #49371 that solves the same problem.

**Pull request closed without merging**